src/test/components/Viewport.test.js

```diff
@@ -0,0 +1,601 @@
+            marginLeft,
+            marginRight,
+          });
+          const viewport = scrollAndGetViewport({
+            deltaY: 100,
+            shiftKey: true,
+            clientX: LEFT + innerComponentWidth * 0.5 + marginLeft,
+          });
+          expect(viewport).toMatchObject({
+            viewportLeft: 0,
+            viewportRight: 1,
+          });
+        });
+      });
+    }
+  });
+
+  describe('dragging around', function() {
+    const middleX = COMPONENT_WIDTH * 0.5;
+    const middleY = COMPONENT_HEIGHT * 0.5;
+
+    it('can click and drag down', function() {
+      const { getChartProps, clickAndDrag } = setup();
+
+      expect(getChartProps().viewport).toMatchObject({
+        viewportLeft: 0,
+        viewportRight: 1,
+        viewportTop: 0,
+        viewportBottom: COMPONENT_HEIGHT,
+      });
+      clickAndDrag(middleX, middleY, middleX, middleY - 50);
+      expect(getChartProps().viewport).toMatchObject({
+        viewportLeft: 0,
+        viewportRight: 1,
+        viewportTop: 50,
+        viewportBottom: COMPONENT_HEIGHT + 50,
+      });
+    });
+
+    it('can click and drag up', function() {
+      const { getChartProps, clickAndDrag } = setup();
+      const middleX = COMPONENT_WIDTH * 0.5;
+      const middleY = COMPONENT_HEIGHT * 0.5;
+
+      clickAndDrag(middleX, middleY, middleX, middleY - 100);
+      expect(getChartProps().viewport).toMatchObject({
+        viewportLeft: 0,
+        viewportRight: 1,
+        viewportTop: 100,
+        viewportBottom: COMPONENT_HEIGHT + 100,
+      });
+      clickAndDrag(middleX, middleY, middleX, middleY + 50);
+      expect(getChartProps().viewport).toMatchObject({
+        viewportLeft: 0,
+        viewportRight: 1,
+        viewportTop: 50,
+        viewportBottom: COMPONENT_HEIGHT + 50,
+      });
+    });
+
+    it('will not drag beyond the top of the window', function() {
+      const { getChartProps, clickAndDrag } = setup();
+      const middleX = COMPONENT_WIDTH * 0.5;
+      const middleY = COMPONENT_HEIGHT * 0.5;
+
+      clickAndDrag(middleX, middleY, middleX, middleY + 50);
+      expect(getChartProps().viewport).toMatchObject({
+        viewportLeft: 0,
+        viewportRight: 1,
+        viewportTop: 0,
+        viewportBottom: COMPONENT_HEIGHT,
+      });
+    });
+
+    it('will not drag beyond the bottom of the window', function() {
+      const { getChartProps, clickAndDrag } = setup();
+      const middleX = COMPONENT_WIDTH * 0.5;
+      const middleY = COMPONENT_HEIGHT * 0.5;
+
+      clickAndDrag(
+        middleX,
+        middleY,
+        middleX,
+        middleY - MAX_VIEWPORT_HEIGHT * 2
+      );
+      expect(getChartProps().viewport).toMatchObject({
+        viewportLeft: 0,
+        viewportRight: 1,
+        viewportTop: MAX_VIEWPORT_HEIGHT - COMPONENT_HEIGHT,
+        viewportBottom: MAX_VIEWPORT_HEIGHT,
+      });
+    });
+
+    it('will anchor small content at the top', function() {
```

Comment:
it's not especially clear what we mean by "anchor" and " small content". Isn't it rather the case of a small view, or rather a small window ?